### PR TITLE
[opt](s3) auto retry when meeting 429 error

### PR DESCRIPTION
**Pull request by morningman**
## Proposed changes

Sometime the s3 sdk will return error like:
```
QpsLimitExceeded Unable to parse ExceptionName: QpsLimitExceeded Message: Please reduce your request rate. code=429 type=100, request_id=66516C288EC49
```
We should slowdown the request rate by sleeping for a while.

- Add 2 new BE config

	- `s3_read_base_wait_time_ms` and `s3_read_max_wait_time_ms`

		When meet s3 429 error, the "get" request will
		sleep `s3_read_base_wait_time_ms (*1, *2, *3, *4)` ms get try again.
		The max sleep time is s3_read_max_wait_time_ms
		and the max retry time is max_s3_client_retry
		
- Add more metrics for s3 file reader

	- `s3_file_reader_too_many_request`: counter of 429 error.
	- `s3_file_reader_s3_get_request`: the QPS of s3 get request.

	- `TotalGetRequest`: Get request counter in profile
	- `TooManyRequestErr`: 429 error counter in profile
	- `TooManyRequestSleepTime`: Sum of sleep time after 429 error in profile
	- `TotalBytesRead`: Total bytes read from s3 in profile

image

### CODE REVIEW

**Comment by morningman** [comment]
run buildall

**Review by kaka11chen — DISMISSED** [review]
LGTM

**Comment by AshinGau** [comment]
LGTM

**Review by gavinchou — DISMISSED** [review]

File: be/src/io/fs/s3_file_reader.cpp [hunk 1]
```diff
@@ -44,13 +45,18 @@ bvar::Adder<uint64_t> s3_file_reader_read_counter("s3_file_reader", "read_at");
 bvar::Adder<uint64_t> s3_file_reader_total("s3_file_reader", "total_num");
 bvar::Adder<uint64_t> s3_bytes_read_total("s3_file_reader", "bytes_read");
 bvar::Adder<uint64_t> s3_file_being_read("s3_file_reader", "file_being_read");
+bvar::Adder<uint64_t> s3_file_reader_too_many_request_counter("s3_file_reader", "too_many_request");
 bvar::LatencyRecorder s3_bytes_per_read("s3_file_reader", "bytes_per_read"); // also QPS
 bvar::PerSecond<bvar::Adder<uint64_t>> s3_read_througthput("s3_file_reader", "s3_read_throughput",
                                                            &s3_bytes_read_total);
+// Although we can get QPS from s3_bytes_per_read, but s3_bytes_per_read only
+// record successfull request, and s3_get_request_qps will record all request.
+bvar::PerSecond<bvar::Adder<uint64_t>> s3_get_request_qps("s3_file_reader", "s3_get_request",
```
Comment (ByteYue) [inline be/src/io/fs/s3_file_reader.cpp:54]: s3_bvar::s3_get_latency can get qps/P99 latency altogether. It seems no need to add this bvar

**Review by ByteYue — CHANGES_REQUESTED** [review]

**Comment by morningman** [comment]
run buildall

File: be/src/io/fs/s3_file_reader.cpp (hunk 1 above)
Comment (morningman) [reply be/src/io/fs/s3_file_reader.cpp:55]: `s3_get_latency` can only save "success" request.
`s3_get_request_qps` saves both "success" and "failed request"

File: be/src/io/fs/s3_file_reader.cpp [hunk 2]
```diff
@@ -113,23 +120,85 @@ Status S3FileReader::read_at_impl(size_t offset, Slice result, size_t* bytes_rea
...
+    int retry_count = 0;
+    const int base_wait_time = config::s3_read_base_wait_time_ms; // Base wait time in milliseconds
+    const int max_wait_time = config::s3_read_max_wait_time_ms; // Maximum wait time in milliseconds
+    const int max_retries = config::max_s3_client_retry; // wait 1s, 2s, 4s, 8s for each backoff
+
+    int total_sleep_time = 0;
+    while (retry_count <= max_retries) {
+        s3_file_reader_read_counter << 1;
+        // clang-format off
+        auto resp = client->get_object( { .bucket = _bucket, .key = _key, },
```
Comment (ByteYue) [inline be/src/io/fs/s3_file_reader.cpp:145]:
```suggestion
        {
        SCOPED_BVAR_LATENCY(s3_bvar::s3_get_latency);
        auto resp = client->get_object( { .bucket = _bucket, .key = _key, },
        }
```

**Review by ByteYue — APPROVED** [review]

**Review by AshinGau — APPROVED** [review]

**Review by kaka11chen — APPROVED** [review]
LGTM

**Review by gavinchou — APPROVED** [review]

**Pull request merged**